README.md
# Read carefully before you continue!
This is just a copy of **[discord-voice](https://github.com/discord-voice/discord-voice)** and has been modified to suit our needs, You should go there because it was the official version of this copy. You will not get any support here.

# What is Voice Utility?

> Voice Utility is a powerful [Node.js](https://nodejs.org/) module that allows you to easily track the user's voice time and levels!

# Features

-   **✨ Easy to use!**
-   **📁 Support for all databases! (default is json)**
-   **⚙️ Very customizable! (ignored channels, ignored members, ignored permissions, xp amount to add, voice time to add etc...)**
-   **🚀 Super powerful: createUser, createConfig, removeUser, removeConfig, updateUser and updateConfig!**
-   **🕸️ Support for shards!**
-   **and much more!**

# Installation

```bash
npm install github:tjs-pro-gg/voice-utility
```

# Usage of the module

```js
const { Client, Intents } = require("discord.js"),
    client = new Client({
        intents: [Intents.FLAGS.GUILD_VOICE_STATES, Intents.FLAGS.GUILDS, Intents.FLAGS.GUILD_MESSAGES] // The GUILD_VOICE_STATES and GUILDS intents are required for voice-utility to function.
    }),
    settings = {
        prefix: "!",
        token: "Your Discord Bot Token"
    };

// Requires Manager from voice-utility
const { VoiceManager } = require("voice-utility");
// Create a new instance of the manager class
const manager = new VoiceManager(client, {
    userStorage: "./users.json",
    configStorage: "./configs.json",
    checkMembersEvery: 5000,
    default: {
        trackBots: false,
        trackAllChannels: true
    }
});
// We now have a voiceManager property to access the manager everywhere!
client.voiceManager = manager;
```

After that, user's who are in the voice channel's that the bot has cached will be checked. You can pass an options object to customize the config. 

# Create an user

```js
client.on("messageCreate", (message) => {
    const args = message.content.slice(settings.prefix.length).trim().split(/ +/g);
    const command = args.shift().toLowerCase();

    if (command === "create-user") {
        client.voiceManager.createUser(message.author.id, message.guild.id, {
            levelingData: {
                xp: 0,
                level: 0
            },
            coindata: {
                point: 0,
                coin: 0
            }
            // The user will have 0 xp, 0 level, 0 point and 0 coin.
        });
    }
});
```

This allow's you create a user in the database if the user is not already present in the database. You can pass an options object to customize the user's data.

# Create a config

```js
client.on("messageCreate", (message) => {
    const args = message.content.slice(settings.prefix.length).trim().split(/ +/g);
    const command = args.shift().toLowerCase();

    if (command === "create-config") {
        client.voiceManager.createConfig(message.guild.id, {
            trackBots: false, // If the user is a bot it will not be tracked.
            trackAllChannels: true, // All of the channels in the guild will be tracked.
            exemptChannels: () => false, // The user will not be tracked in these channels. (This is a function).
            channelIds: [], // The channel ids to track. (If trackAllChannels is true, this is ignored)
            exemptPermissions: [], // The user permissions to not track.
            exemptMembers: () => false, // The user will not be tracked. (This is a function).
            trackMute: true, // It will track users if they are muted aswell.
            trackDeaf: true, // It will track users if they are deafen aswell.
            minUserCountToParticipate: 0, // The min amount of users to be in a channel to be tracked.
            maxUserCountToParticipate: 0, // The max amount of users to be in a channel to be tracked.
            minPointToParticipate: 0, // The min amount of point needed to be tracked.
            minCoinToParticipate: 0, // The min coin needed to be tracked.
            minXpToParticipate: 0, // The min amount of xp needed to be tracked.
            minLevelToParticipate: 0, // The min level needed to be tracked.
            maxPointToParticipate: 0, // The max amount of point needed to be tracked.
            maxCoinToParticipate: 0, // The max amount of coin needed to be tracked.
            maxXpToParticipate: 0, // The max amount of xp needed to be tracked.
            maxLevelToParticipate: 0, // The max level needed to be tracked.
            pointAmountToAdd: () => Math.floor(Math.random() * 5) + 1, // The amount of xp to add to the user (This is a function).
            xpAmountToAdd: () => Math.floor(Math.random() * 10) + 1, // The amount of xp to add to the user (This is a function).
            voiceTimeToAdd: () => 1000, // The amount of time in ms to add to the user (This is a function).
            voiceTimeTrackingEnabled: true, // Whether the voiceTimeTracking module is enabled.
            levelingTrackingEnabled: true, // Whether the levelingTracking module is enabled.
            coinTrackingEnabled: true // Whether the coinTracking module is enabled.
        });
    }
});
```

This allow's you create a config in the database if the config is not already present in the database. You can pass an options object to customize the config's data.

# Remove an user

```js
client.on("messageCreate", (message) => {
    const args = message.content.slice(settings.prefix.length).trim().split(/ +/g);
    const command = args.shift().toLowerCase();

    if (command === "remove-user") {
        client.voiceManager.removeUser(message.author.id, message.guild.id); // Removes the user from the database and the cache.
    }
});
```

# Remove a config

```js
client.on("messageCreate", (message) => {
    const args = message.content.slice(settings.prefix.length).trim().split(/ +/g);
    const command = args.shift().toLowerCase();

    if (command === "remove-config") {
        client.voiceManager.removeConfig(message.guild.id); // Removes the config from the database and the cache.
    }
});
```

# Updating an user

```js
client.on("messageCreate", (message) => {
    const args = message.content.slice(settings.prefix.length).trim().split(/ +/g);
    const command = args.shift().toLowerCase();

    if (command === "edit-user") {
        client.voiceManager.updateUser(message.author.id, message.guild.id, {
            newVoiceTime: {
                channels: [],
                total: 0
            } // The new voice time user will have.
        });
    }
});
```

This allow's you edit a user's data. You need to pass an options object to edit the user's data.

# Updating a config

```js
client.on("messageCreate", (message) => {
    const args = message.content.slice(settings.prefix.length).trim().split(/ +/g);
    const command = args.shift().toLowerCase();

    if (command === "edit-config") {
        client.voiceManager.updateConfig(message.guild.id, {
            newTrackBots: true // The module will now track bot user's voice time aswell.
        });
    }
});
```

This allow's you edit a config's data. You need to pass an options object to edit the config's data.

# Fetch users

```js
// A list of all the users
const allUsers = client.voiceManager.users; // [ {User}, {User} ]

// A list of all the users on the server with ID "1909282092"
const onServer = client.voiceManager.users.filter((u) => u.guildId === "1909282092");

// The user on the server with Id "1909282092" and the user Id "1234567890"
const user = client.voiceManager.users.filter((u) => u.guildId === "1909282092" && u.userId === "1234567890");
```

# Fetch configs

```js
// A list of all the configs
const allConfigs = client.voiceManager.configs; // [ {Config}, {Config} ]

// The config of the guild with Id "1909282092"
const config = client.voiceManager.configs.filter((c) => c.guildId === "1909282092");
```

# Exempt Channels

```js
client.voiceManager.updateConfig(message.guild.id, {
    // The channel will not be tracked if it's name is "private"
    exemptChannels: (channel) => channel.name === "private")
});
```

⚠️ Note: If the function should be customizable

```js
const channelName = "private";
client.voiceManager.updateConfig(message.guild.id, {
    // The channel won't be tracked if it's name is equal to the value which is assigned to "channelName"
    exemptChannels: new Function("channel", `return channel.name === \'${channelName}\'`)
});
```

# Exempt Members

```js
client.voiceManager.updateConfig(message.guild.id, {

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.1.13](https://github.com/tjs-pro-gg/voice-utility/compare/v1.1.12...v1.1.13) (2022-05-02)


### Bug Fixes

* fix incorrect event name ([ead9ae8](https://github.com/tjs-pro-gg/voice-utility/commit/ead9ae81bd8638660164450962d18b7b0344ffb3))

### [1.1.12](https://github.com/tjs-pro-gg/voice-utility/compare/v1.1.11...v1.1.12) (2022-04-29)


### Bug Fixes

* this should be optional type ([615f92a](https://github.com/tjs-pro-gg/voice-utility/commit/615f92ad47fdff3da477dc4dab883c57d70c9733))

### [1.1.11](https://github.com/tjs-pro-gg/voice-utility/compare/v1.1.10...v1.1.11) (2022-04-29)


### Bug Fixes

* coinData does not exist ([cc8a9fe](https://github.com/tjs-pro-gg/voice-utility/commit/cc8a9fe54ff8ea506fff007f152fa423be35e692))
* coinData is not exist ([930f656](https://github.com/tjs-pro-gg/voice-utility/commit/930f65679e513f4f4389d753d5b1648d9c4385c2))
* fix the user coin not writed to database ([69cadb7](https://github.com/tjs-pro-gg/voice-utility/commit/69cadb742962e2e5a9732fd999e1b9a5e61f44e6))
* reverting back types ([bab7ba8](https://github.com/tjs-pro-gg/voice-utility/commit/bab7ba8633b97a74780e00a70182c9e648612fc1))
* valid types for ConfigOptions ([025886f](https://github.com/tjs-pro-gg/voice-utility/commit/025886f689596ab6789ec48b9065896c731f60dc))

### [1.1.10](https://github.com/tjs-pro-gg/voice-utility/compare/v1.1.9...v1.1.10) (2022-04-27)


### Bug Fixes

* fix incorrect types ([c57733a](https://github.com/tjs-pro-gg/voice-utility/commit/c57733ab7a712786d91876e99ce2164090843773))
* fix this type ([5dd6851](https://github.com/tjs-pro-gg/voice-utility/commit/5dd6851c728117e3924c3e3c20e2827f03161d16))
* try this types ([5b50eb1](https://github.com/tjs-pro-gg/voice-utility/commit/5b50eb1810c017d9de65e997b703c96bf0491121))


### Tests

* try this types ([6aec9cf](https://github.com/tjs-pro-gg/voice-utility/commit/6aec9cfdafab9d09ffb489d2cf66c88e27e61f40))
* try this types ([f948be0](https://github.com/tjs-pro-gg/voice-utility/commit/f948be0a833db92559bbef23cffa1c0d95ed8847))

### [1.1.9](https://github.com/tjs-pro-gg/voice-utility/compare/v1.1.8...v1.1.9) (2022-04-26)


### Bug Fixes

* fix incorrect tipe for channelId ([37ef06b](https://github.com/tjs-pro-gg/voice-utility/commit/37ef06b8463a841566f79f586611d6511efecc24))


### Enhancements

* change to npm from yarn ([957ffd2](https://github.com/tjs-pro-gg/voice-utility/commit/957ffd269afc8ecf3e727e93895f3833a1122f17))
* **deps:** pin dependency standard-version to 9.3.2 ([#5](https://github.com/tjs-pro-gg/voice-utility/issues/5)) ([6c67ef4](https://github.com/tjs-pro-gg/voice-utility/commit/6c67ef4ac623ef0271a13c0b912b886948a16806))

### [1.1.8](https://github.com/tjs-pro-gg/voice-utility/compare/v1.1.7...v1.1.8) (2022-04-26)


### Bug Fixes

* fix incorrect type of channels ([660a0b9](https://github.com/tjs-pro-gg/voice-utility/commit/660a0b928fe9e05b2661d81388c6a773726ac730))

### [1.1.7](https://github.com/tjs-pro-gg/voice-utility/compare/v1.1.6...v1.1.7) (2022-04-24)


### Bug Fixes

* incorrect module ([b61cedb](https://github.com/tjs-pro-gg/voice-utility/commit/b61cedb9d8638487b7e73580284202f090e68a4f))
* typings ([4fcdd8d](https://github.com/tjs-pro-gg/voice-utility/commit/4fcdd8d8685cbf8898a62e8a45c2225af3dfda72))

### [1.1.6](https://github.com/tjs-pro-gg/voice-utility/compare/v1.1.5...v1.1.6) (2022-04-24)

### [1.1.5](https://github.com/tjs-pro-gg/voice-utility/compare/v1.1.4...v1.1.5) (2022-04-24)

### [1.1.4](https://github.com/tjs-pro-gg/voice-utility/compare/v1.1.3...v1.1.4) (2022-04-24)

### [1.1.3](https://github.com/tjs-pro-gg/voice-utility/compare/v1.1.2...v1.1.3) (2022-04-24)

### [1.1.2](https://github.com/tjs-pro-gg/voice-utility/compare/v1.1.1...v1.1.2) (2022-04-24)


### Bug Fixes

* correct module ([09ea93a](https://github.com/tjs-pro-gg/voice-utility/commit/09ea93abbfadab99c12e2079bfded2a34614a661))

### [1.1.1](https://github.com/tjs-pro-gg/voice-utility/compare/v1.1.0...v1.1.1) (2022-04-24)


### Enhancements


src/lib/Manager.js
const { EventEmitter } = require("events");
const merge = require("deepmerge");
const { writeFile, readFile, access } = require('fs/promises');
const serialize = require("serialize-javascript");
const lodash = require("lodash");
const { defaultVoiceManagerOptions, defaultUserOptions, defaultConfigOptions, VoiceManagerOptions, UserOptions, ConfigOptions, UserData, ConfigData, UserEditOptions, ConfigEditOptions } = require("./Constants.js");
const Config = require("./Config.js");
const User = require("./User.js");

/**
 * Voice Manager
 * @example
 * // Requires Manager from voice-utility
 * const { VoiceManager } = require("voice-utility");
 * // Create a new instance of the manager class
 * const manager = new VoiceManager(client, {
 * userStorage: "./users.json",
 * configStorage: "./configs.json",
 * checkMembersEvery: 5000,
 * default: {
 *   trackBots: false,
 *   trackAllChannels: true,
 * },
 * });
 * // We now have a voiceManager property to access the manager everywhere!
 * client.voiceManager = manager;
 */
class VoiceManager extends EventEmitter {
    /**
     * @param {Client} client The Discord Client
     * @param {VoiceManagerOptions} options The manager options
     */
    constructor(client, options, init = true) {
        super();
        if (!client?.options) throw new Error(`Client is a required option. (val=${client})`);
        /**
         * The Discord Client
         * @type {Client}
         */
        this.client = client;
        /**
         * Whether the manager is ready
         * @type {Boolean}
         */
        this.ready = false;
        /**
         * The user's managed by this manager
         * @type {User[]}
         */
        this.users = [];
        /**
         * The config's managed by this manager
         * @type {Config[]}
         */
        this.configs = [];
        /**
         * The manager options
         * @type {VoiceManagerOptions}
         */
        this.options = merge(defaultVoiceManagerOptions, options);
        if (init) this._init();
    }
    /**
     * Creates a new user in the database
     *
     * @param {Snowflake} userId The id of the user
     * @param {Snowflake} guildId The id of the user's guild
     * @param {UserOptions} options The options for the user
     *
     * @returns {Promise<User>}
     *
     * @example
     * manager.createUser(message.author.id, message.guild.id, {
     *      levelingData: {
     *        xp: 0,
     *        level: 0
     *      },
     *      coinData {
     *        point: 0,
     *        coin: 0
     *      }
     *      // The user will have 0 xp, 0 level, 0 point and 0 coin.
     * });
     */
    createUser(userId, guildId, options) {
        return new Promise(async (resolve, reject) => {
            if (!this.ready) {
                return reject("The manager is not ready yet.");
            }
            options = options && typeof options === "object" ? merge(defaultUserOptions, options) : defaultUserOptions;
            if (!userId) {
                return reject(`userId is not a valid user. (val=${userId})`);
            }
            if (!guildId) {
                return reject(`guildId is not a valid guild. (val=${guildId})`);
            }
            const user = new User(this, {
                userId: userId,
                guildId: guildId,
                data: options
            });
            this.users.push(user);
            await this.saveUser(userId, guildId, user.data);
            resolve(user);
        });
    }
    /**
     * Creates a new config in the database
     *
     * @param {Snowflake} guildId The id of the config's guild
     * @param {ConfigOptions} options The options for config
     *
     * @returns {Promise<Config>}
     *
     * @example
     * manager.createConfig(message.guild.id, {
     *      trackBots: false, // If the user is a bot it will not be tracked.
     *      trackAllChannels: true, // All of the channels in the guild will be tracked.
     *      exemptChannels: () => false, // The user will not be tracked in these channels. (This is a function).
     *      channelIds: [], // The channel ids to track. (If trackAllChannels is true, this is ignored)
     *      exemptPermissions: [], // The user permissions to not track.
     *      exemptMembers: () => false, // The user will not be tracked. (This is a function).
     *      trackMute: true, // It will track users if they are muted aswell.
     *      trackDeaf: true, // It will track users if they are deafen aswell.
     *      minUserCountToParticipate: 0, // The min amount of users to be in a channel to be tracked.
     *      maxUserCountToParticipate: 0, // The max amount of users to be in a channel to be tracked.
     *      minPointToParticipate: 0, // The min amount of point needed to be tracked.
     *      minCoinToParticipate: 0, // The min coin needed to be tracked.
     *      minXpToParticipate: 0, // The min amount of xp needed to be tracked.
     *      minLevelToParticipate: 0, // The min level needed to be tracked.
     *      maxPointToParticipate: 0, // The max amount of point needed to be tracked.
     *      maxCoinToParticipate: 0, // The max coin needed to be tracked.
     *      maxXpToParticipate: 0, // The max amount of xp needed to be tracked.
     *      maxLevelToParticipate: 0, // The max level needed to be tracked.
     *      xpAmountToAdd: () => Math.floor(Math.random() * 10) + 1, // The amount of xp to add to the user (This is a function).
     *      voiceTimeToAdd: () => 1000, // The amount of time in ms to add to the user (This is a function).
     *      voiceTimeTrackingEnabled: true, // Whether the voiceTimeTracking module is enabled.
     *      levelingTrackingEnabled: true, // Whether the levelingTracking module is enabled.
     *      coinTrackingEnabled: true, // Whether the coinTracking module is enabled.
     *      pointMultiplier: () => 0.1, // This will set point multiplier to 0.1 (This is a function).
     *      levelMultiplier: () => 0.1, // This will set level multiplier to 0.1 (This is a function).
     * });
     */
    createConfig(guildId, options) {
        return new Promise(async (resolve, reject) => {
            if (!this.ready) {
                return reject("The manager is not ready yet.");
            }
            options = options && typeof options === "object" ? merge(defaultConfigOptions, options) : defaultConfigOptions;
            if (!guildId) {
                return reject(`guildId is not a valid guild. (val=${guildId})`);
            }
            const config = new Config(this, {
                guildId: guildId,
                data: options
            });
            this.configs.push(config);
            await this.saveConfig(guildId, config.data);
            resolve(config);
        });
    }
    /**
     * Remove's the user from the database
     *
     * @param {Snowflake} userId The id of the user
     * @param {Snowflake} guildId The id of the user's guild
     *
     * @returns {Promise<void>}
     *
     * @example
     * manager.removeUser(message.author.id, message.guild.id);
     */
    removeUser(userId, guildId) {
        return new Promise(async (resolve, reject) => {
            const user = this.users.find((u) => u.guildId === guildId && u.userId === userId);
            if (!user) {
                return reject("No user found with Id " + userId + " in guild with Id" + guildId + ".");
            }
            this.users = this.users.filter(
                (d) =>
                    d !==
                    {
                        userId: userId,
                        guildId: guildId,
                        data: user.data.data
                    }
            );
            await this.deleteUser(userId, guildId);
            resolve();
        });
    }
    /**
     * Remove's the config from the database
     *
     * @param {Snowflake} guildId The id of the config's guild
     *
     * @returns {Promise<void>}
     *
     * @example
     * manager.removeConfig(message.guild.id);
     */
    removeConfig(guildId) {
        return new Promise(async (resolve, reject) => {
            const config = this.configs.find((c) => c.guildId === guildId);
            if (!config) {
                return reject("No config found for guild with Id " + guildId + ".");
            }
            this.configs = this.configs.filter((c) => c.guildId !== guildId);
            await this.deleteConfig(guildId);
            resolve();
        });
    }
    /**
     * Edits a user. The modifications will be applicated when the user will be updated.
     * @param {Snowflake} userId The id of the user
     * @param {Snowflake} guildId The id of the user's guild
     * @param {UserEditOptions} options The edit options
     * @returns {Promise<User>}
     *
     * @example
     * manager.updateUser('122925169588043776','815261972450115585', {
     *      newVoiceTime: {
     *      channels: [],
     *      total: 0,
     *      }, // The new voice time user will have.
     * });
     */
    updateUser(userId, guildId, options = {}) {
        return new Promise(async (resolve, reject) => {
            const user = this.users.find((u) => u.guildId === guildId && u.userId === userId);
            if (!user) {
                return reject("No user found with Id " + userId + " in guild with Id" + guildId + ".");
            }
            user.edit(options).then(resolve).catch(reject);
        });
    }
    /**
     * Edits a config.
     * @param {Snowflake} guildId The id of the user's guild
     * @param {ConfigEditOptions} options The edit options
     * @returns {Promise<Config>}
     *
     * @example
     * manager.updateConfig('815261972450115585', {
     *      newTrackBots: true, // The module will now track bot user's voice time aswell.
     * });
     */
    updateConfig(guildId, options = {}) {
        return new Promise(async (resolve, reject) => {
            const config = this.configs.find((c) => c.guildId === guildId);
            if (!config) {
                return reject("No config found for guild with Id " + guildId + ".");
            }
            config.edit(options).then(resolve).catch(reject);
        });
    }
    /**
     * Delete a user from the database
     * @param {Snowflake} userId The id of the user
     * @param {Snowflake} guildId The id of the user's guild
     * @returns {Promise<void>}
     * @ignore
     */
    async deleteUser(userId, guildId) {
        await writeFile(this.options.userStorage,
        JSON.stringify(this.users.map((user) => user.data), (_, v) => typeof v === 'bigint' ? serialize(v) : v), 
        'utf-8'
        );
        this.refreshUserStorage();
        return;
    }
    /**
     * Delete a config from the database
     * @param {Snowflake} guildId The id of the config's guild
     * @returns {Promise<void>}
     * @ignore
     */
    async deleteConfig(guildId) {
        await writeFile(
            this.options.configStorage,
            JSON.stringify(
                this.configs.map((config) => config.data),
                (_, v) => (typeof v === "bigint" ? serialize(v) : v)
            ),
            'utf-8'
        );
        this.refreshConfigStorage();
        return;
    }
    /**
     * Refresh the user cache to support shards.
     * @ignore
     */
    async refreshUserStorage() {
        return true;
    }
    /**
     * Refresh the config cache to support shards.
     * @ignore
     */
    async refreshConfigStorage() {
        return true;
    }
    /**
     * Edit the user in the database
     * @ignore
     * @param {Snowflake} userId The id of the user
     * @param {Snowflake} guildId The id of the user's guild
     * @param {UserData} userData The user data to save
     */
    async editUser(_userId, _guildId, _userData) {
        await writeFile(
            this.options.userStorage,
            JSON.stringify(
                this.users.map((user) => user.data),
                (_, v) => (typeof v === "bigint" ? serialize(v) : v)
            ),
            'utf-8'
        );
        this.refreshUserStorage();
        return;
    }
    /**
     * Edit the config in the database
     * @ignore
     * @param {Snowflake} guildId The id of the config's guild
     * @param {ConfigData} ConfigData The config data to save
     */
    async editConfig(_guildId, _configData) {
        await writeFile(
            this.options.configStorage,
            JSON.stringify(
                this.configs.map((config) => config.data),
                (_, v) => (typeof v === "bigint" ? serialize(v) : v)
            ),
            'utf-8'
        );
        this.refreshConfigStorage();
        return;
    }
    /**
     * Save the user in the database
     * @ignore
     * @param {Snowflake} userId The id of the user
     * @param {Snowflake} guildId The id of the user's guild
     * @param {UserData} userData The user data to save
     */
    async saveUser(userId, guildId, userData) {
        await writeFile(
            this.options.userStorage,
            JSON.stringify(
                this.users.map((user) => user.data),
                (_, v) => (typeof v === "bigint" ? serialize(v) : v)
            ),
            'utf-8'
        );
        this.refreshUserStorage();
        return;
    }
    /**
     * Save the config in the database
     * @ignore
     * @param {Snowflake} guildId The id of the config's guild
     * @param {ConfigData} configData The config data to save
     */
    async saveConfig(guildId, configData) {
        await writeFile(
            this.options.configStorage,
            JSON.stringify(
                this.configs.map((config) => config.data),
                (_, v) => (typeof v === "bigint" ? serialize(v) : v)
            ),
            'utf-8'
        );
        this.refreshConfigStorage();
        return;
    }
    /**
     * Gets the user's from the storage file, or create it
     * @ignore
     * @returns {Promise<UserData[]>}
     */
    async getAllUsers() {
        const storageExists = await access(this.options.userStorage)
            .then(() => true)
            .catch(() => false);
        if (!storageExists) {
            await writeFile(this.options.userStorage, '[]', 'utf-8');
            return [];
        } else {
            const storageContent = await readFile(this.options.userStorage, (_, v) => (typeof v === "string" && /BigInt\("(-?\d+)"\)/.test(v) ? eval(v) : v));
            try {
                const users = await JSON.parse(storageContent.toString());
                if (Array.isArray(users)) {
                    return users;
                } else {
                    console.log(storageContent, users);
                    throw new SyntaxError("The storage file is not properly formatted (users is not an array).");
                }
            } catch (err) {
                if (err.message === "Unexpected end of JSON input") {
                    throw new SyntaxError("The storage file is not properly formatted (Unexpected end of JSON input).");
                } else throw err;
            }
        }
    }
    /**
     * Gets the config's from the storage file, or create it
     * @ignore
     * @returns {Promise<ConfigData[]>}
     */
    async getAllConfigs() {
        const storageExists = await access(this.options.configStorage)
            .then(() => true)
            .catch(() => false);
        if (!storageExists) {
            await writeFile(this.options.configStorage, '[]', 'utf-8');
            return [];
        } else {
            const storageContent = await readFile(this.options.configStorage, (_, v) => (typeof v === "string" && /BigInt\("(-?\d+)"\)/.test(v) ? eval(v) : v));
            try {
                const configs = await JSON.parse(storageContent.toString());
                if (Array.isArray(configs)) {
                    return configs;
                } else {
                    console.log(storageContent, configs);
                    throw new SyntaxError("The storage file is not properly formatted (configs is not an array).");
                }
            } catch (err) {
                if (err.message === "Unexpected end of JSON input") {
                    throw new SyntaxError("The storage file is not properly formatted (Unexpected end of JSON input).");
                } else throw err;
            }
        }
    }
    /**
     * Checks each user and update it if needed
     * @ignore
     * @private
     */
    _checkUsers() {
        if (this.users.length <= 0) return;
        this.users.forEach(async (user) => {
            if (user.member && user.channel) {
                let config = this.configs.find((g) => g.guildId === user.guildId);
                if (!config) {
                    config = await this.createConfig(user.guildId);
                }
                if (!((await config.checkMember(user.member)) && (await config.checkChannel(user.channel)))) return;
                const oldUser = lodash._.cloneDeep(user);
                if (config.voiceTimeTrackingEnabled) {
                    let previousVoiceTime;
                    user.voiceTime.channels.length <= 0
                        ? (previousVoiceTime = {
                              channelId: user.channel.id,
                              voiceTime: 0
                          })
                        : user.voiceTime.channels.find((chn) => chn.channelId === user.channel.id)
                        ? (previousVoiceTime = user.voiceTime.channels.find((chn) => chn.channelId === user.channel.id))
                        : (previousVoiceTime = {
                              channelId: user.channel.id,
                              voiceTime: 0
                          });
                    let index = user.voiceTime.channels.indexOf(previousVoiceTime);
                    previousVoiceTime.voiceTime += await config.voiceTimeToAdd();
                    if (index === -1) user.voiceTime.channels.push(previousVoiceTime);
                    else user.voiceTime.channels[index] = previousVoiceTime;
                    user.voiceTime.total = user.voiceTime.channels.reduce(function (sum, data) {
                        return sum + data.voiceTime;
                    }, 0);
                    /**
                     * Emitted when voice time is added to the user.
                     * @event VoiceManager#userVoiceTimeAdd
                     * @param {User} oldUser The user before the update
                     * @param {User} newUser The user after the update
                     *
                     */
                    this.emit("userVoiceTimeAdd", oldUser, user);
                }
                if (config.levelingTrackingEnabled) {
                    user.levelingData.xp += await config.xpAmountToAdd();
                    user.levelingData.level = Math.floor((await config.levelMultiplier()) * Math.sqrt(user.levelingData.xp));
                    /**
                     * Emitted when xp is added to the user.
                     * @event VoiceManager#userXpAdd
                     * @param {User} oldUser The user before the update
                     * @param {User} newUser The user after the update
                     *
                     */
                    this.emit("userXpAdd", oldUser, user);
                    if (user.levelingData.level > oldUser.levelingData.level) {
                        /**
                         * Emitted when the user levels up.
                         * @event VoiceManager#userLevelUp
                         * @param {User} oldUser The user before the update
                         * @param {User} newUser The user after the update
                         *
                         */
                        this.emit("userLevelUp", oldUser, user);
                    }
                }

                if (config.coinTrackingEnabled) {
                    user.coinData.point += await config.pointAmountToAdd();
                    user.coinData.coin = Math.floor((await config.pointMultiplier()) * Math.sqrt(user.coinData.point));
                    /**
                     * Emitted when point is added to the user.
                     * @event VoiceManager#userPointAdd
                     * @param {User} oldUser The user before the update
                     * @param {User} newUser The user after the update
                     *
                     */
                    this.emit("userPointAdd", oldUser, user);
                    if (user.coinData.coin > oldUser.coinData.coin) {
                        /**
                         * Emitted when the user coins add.
                         * @event VoiceManager#userCoinAdd
                         * @param {User} oldUser The user before the update
                         * @param {User} newUser The user after the update
                         *
                         */
                        this.emit("userCoinAdd", oldUser, user);
                    }
                }
                await this.editUser(user.userId, user.guildId, user.data);
                return;
            }
        });
    }
    /**
     * Checks the provided user
     * @ignore
     * @private
     */
    async _checkUser(memberAndChannel) {
        let config = this.configs.find((g) => g.guildId === memberAndChannel.member.guild.id);
        if (!config) {
            config = await this.createConfig(memberAndChannel.member.guild.id);
        }
        if (!((await config.checkMember(memberAndChannel.member)) && (await config.checkChannel(memberAndChannel.channel)))) return false;
        else return await this.createUser(memberAndChannel.member.id, memberAndChannel.member.guild.id);
    }
    /**
     * Saves the new user to the storage file
     * @ignore
     * @private
     */
    async _handleVoiceStateUpdate(oldState, newState) {
        if (newState.channel) {
            if (!this.users.find((u) => u.userId === newState.member.id)) {
                let config = this.configs.find((g) => g.guildId === newState.member.guild.id);
                if (!config) {
                    config = await this.createConfig(newState.member.guild.id);
                }
                if (!((await config.checkMember(newState.member)) && (await config.checkChannel(newState.channel)))) return;
                else return await this.createUser(newState.member.id, newState.member.guild.id);
            }
        }
    }
    /**
     * Inits the manager
     * @ignore
     * @private
     */
    async _init() {
        const rawUsers = await this.getAllUsers();
        rawUsers.forEach((user) => {
            this.users.push(new User(this, user));
        });
        const rawConfig = await this.getAllConfigs();
        rawConfig.forEach((config) => {
            this.configs.push(new Config(this, config));
        });
        setInterval(() => {
            if (this.client.readyAt) this._checkUsers.call(this);
        }, this.options.checkMembersEvery);
        this.ready = true;
        this.client.on("voiceStateUpdate", (oldState, newState) => this._handleVoiceStateUpdate(oldState, newState));
    }
}

module.exports = VoiceManager;
